oci/layer/tar_extract.go

```diff
@@ -167,6 +167,31 @@ func (te *tarExtractor) applyMetadata(path string, hdr *tar.Header) error {
 	return te.restoreMetadata(path, hdr)
 }
 
+func (te *tarExtractor) isDirlink(root string, path string, hdr *tar.Header) (bool, error) {
+	link, err := te.fsEval.Readlink(path)
+	if err != nil {
+		return false, errors.Wrap(err, "read existing link")
+	}
+
+	linkPath, err := securejoin.SecureJoinVFS(root, link, te.fsEval)
+	if err != nil {
+		return false, errors.Wrap(err, "sanitize old target")
+	}
+
+	linkInfo, err := te.fsEval.Lstat(linkPath)
```

Comment:
I just realised that this won't handle nested symlinks (a symlink to a symlink to a directory) properly -- we'd need to loop this logic and bail with an `ELOOP`. (I really wish we had `openat(..., AT_THIS_ROOT)` so we didn't have to write this logic in userspace.)